server/market/market.go

```diff
@@ -3,30 +3,251 @@
-// various account package callbacks such as order status updates so that the
-// account package code can keep various data up-to-date, including order
-// status, history, cancellation statistics, etc.
-type Market struct{} // TODO
+	// Communications.
+	orderRouter chan *orderUpdateSignal  // incoming orders
+	orderFeeds  []chan *BookUpdateSignal // outgoing notifications
+
+	// Order handling: book, active epoch queues, order matcher, and swapper.
+	marketInfo    *types.MarketInfo
+	lotSize       uint64
+	book          *book.Book
+	epochDuration uint64
+	epochQueue    *EpochQueue
+	matcher       *matcher.Matcher
+	swapper       *swap.Swapper
+
+	// The asset backends.
+	baseAsset  *asset.Asset
+	quoteAsset *asset.Asset
+
+	// Persistent data storage.
+	storage db.DEXArchivist
+}
+
+func (m *Market) NewBookUpdateChan() <-chan *BookUpdateSignal {
+	bookUpdates := make(chan *BookUpdateSignal)
+	m.orderFeeds = append(m.orderFeeds, bookUpdates)
+	return bookUpdates
+}
+
+func (m *Market) SubmitOrder(ord order.Order) error {
+	return <-m.SubmitOrderAsync(ord)
+}
+
+func (m *Market) SubmitOrderAsync(ord order.Order) <-chan error {
+	sig := newOrderUpdateSignal(ord)
+	m.orderRouter <- sig
+	return sig.errChan
+}
+
+func NewMarket(ctx context.Context, lotSize, epochDuration uint64, baseAsset, quoteAsset *asset.Asset,
+	storage db.DEXArchivist, authMgr swap.AuthManager) (*Market, error) {
+
+	mktInfo, err := types.NewMarketInfoFromSymbols(baseAsset.Symbol, quoteAsset.Symbol, lotSize, epochDuration)
+	if err != nil {
+		return nil, err
+	}
+
+	swapperCfg := &swap.Config{
+		Ctx: ctx,
+		Assets: map[uint32]*asset.Asset{
+			mktInfo.Base:  baseAsset,
+			mktInfo.Quote: quoteAsset,
+		},
+		Storage:          storage,
+		AuthManager:      authMgr,
+		BroadcastTimeout: 10 * time.Second, // TODO: who sets this?
+	}
+
+	return &Market{
+		ctx:           ctx,
+		marketInfo:    mktInfo,
+		lotSize:       lotSize,
+		epochDuration: epochDuration,
+		orderRouter:   make(chan *orderUpdateSignal, 16),
+		book:          book.New(lotSize),
+		matcher:       matcher.New(),
+		swapper:       swap.NewSwapper(swapperCfg),
+		baseAsset:     baseAsset,
+		quoteAsset:    quoteAsset,
+		storage:       storage,
+	}, nil
+}
+
+func (m *Market) start() {
+	m.wg.Add(1)
+	go m.runEpochs()
+}
+
+func (m *Market) notifyNewOrder(order order.Order, action bookUpdateAction) {
+	// send to each receiver
+	for _, s := range m.orderFeeds {
+		select {
+		case <-m.ctx.Done():
+			return
+		case s <- &BookUpdateSignal{action, order}:
+		}
+	}
+}
+
+func (m *Market) runEpochs() {
+	// Startup epoch is the epoch after the epoch one second in the future.
+	dur := int64(m.epochDuration)
+	nextEpochIdx := 1 + (1+time.Now().Unix())%dur
+	nextEpoch := NewEpoch(nextEpochIdx, dur)
+	epochCycle := time.After(time.Until(nextEpoch.Start))
+
+	defer m.wg.Done()
+
+	for {
+		select {
+		case <-m.ctx.Done():
+			m.running = false
+			return
+
+		case <-epochCycle:
```

Comment:
I'm not sure how this will play out, but we may want to trigger the epoch match cycle slightly differently. [There is very little happening](https://github.com/decred/dcrdex/blob/9cb00996839b677a516c3810ec4f5a56287e975b/server/market/orderrouter.go#L160-L181) between the time the order is timestamped and when it is submitted, but the channel could get backed up such that this cycle will trigger while an order for this epoch is still in the channel.

Reply:
I'm glad you're on to this because I agree; I considered that; I don't have the solution.  :)

My current point of view is presently that the epoch should cycle as defined by `time % duration`, but if an order with a server time that would land in that epoch doesn't make it then it can just go in the next epoch.  Since there's no way (that I can see) to make the order processing pipeline not a race without delaying epoch processing until the incoming order pipeline is cleared (to some extent based on incoming order times), we may just have to forfeit any guarantees that an order with a given server time stamp lands in a particular epoch.  This is a touchy issue though as it gets to server censorship and gaming.  

Just the fact that the server sets a server timestamp at its convenience is a bit irksome.

Reply:
Yep. An epoch not based on the signed server timestamp sent to the client wouldn't really work for us. We could reject orders that come in after the epoch cycle starts, but I don't think it's unreasonable to delay matching up to some small pre-defined limit to wait for processing to complete. Processing of the epoch could be triggered by either 1) seeing an order from a new epoch or 2) a final timeout, after which orders would be rejected.

I don't really like using timeouts like that, but I don't see a clear solution here so I'm open to it.

Reply:
Maybe the order router can send an "epoch closed" signal.

Reply:
But I don't love that the order router would have to care about epochs.

Reply:
We could have the Market actually timestamp the order and send the response to the client, instead of the order router. Basically handle [this part](https://github.com/decred/dcrdex/blob/9cb00996839b677a516c3810ec4f5a56287e975b/server/market/orderrouter.go#L185-L199) in Market.